### PR TITLE
Add fullscreen API

### PR DESCRIPTION
**Pull request by MortimerGoro**
- Receive DOM fullscreen state via wpe_view_backend_set_fullscreen
- Dispatch exit/enter fullscreem from libwpe to the client (e.g. WPEView in Webkit)

----

Corresponding WPE WebKit patch: https://bugs.webkit.org/show_bug.cgi?id=227951

### CODE REVIEW

File: src/view-backend.c
Comment (aperezdc) [inline src/view-backend.c:202]: Same here: is it expected that this function would be called only once per view backend? Do we want to allow changing the handler?

We may want to cover the case in which fullscreening is requested by the application using WPE determines that it is not currently possible. I think the callback should have type `bool (*))(void*, bool)`; and if it returns `false`, cancel the fullscreening operation on the WebKit side.

Finally, it is unclear who has to implement a “client” and who has to provide a “handler”, having some documentation about what each does would bee great.

**Review by aperezdc — COMMENTED** [review]
Functionality-wise I think the patch looks fine, though it was a bit hard to follow what's going on and how the API is to be used without looking at the accompanying [patch for WebKit](https://bugs.webkit.org/show_bug.cgi?id=227951) and the [PR for Cog](https://github.com/Igalia/cog/pull/329). Would it be possible to add some reference documentation?

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h:205]: done

File: src/view-backend.c
Comment (MortimerGoro) [reply src/view-backend.c]: I have improved the API with better function names, some comments and implementing the wait for the new exported image we talked about

File: src/view-backend.c
Comment (MortimerGoro) [reply src/view-backend.c:103]: Do you have any example on how to assert that check? I mean do you mean adding a assert/panic or similar.

Checking a similar function (wpe_view_backend_set_input_client) I don't see it having the check or the destroy_notify callback

Comment (aperezdc) [reply src/view-backend.c:103]: We don't have such checks in other functions, but it would be good to have them. I think a plain `assert(!backend->fullscreen_client)` should be enough here 🤔

File: src/view-backend.c
Comment (aperezdc) [inline src/view-backend.c:69]: I'm somewhat amused that this does not cause compiler warnings; we should be using `NULL` here because the members are pointers 🤔. Anyhoo, let's use it:

```suggestion
    backend->fullscreen_client = NULL;
    backend->fullscreen_client_data = NULL;
```

File: src/view-backend.c
Comment (aperezdc) [reply src/view-backend.c:103]: I think you might have forgotten to add the assertion here 😉

File: include/wpe/view-backend.h
Comment (aperezdc) [inline include/wpe/view-backend.h:110]: Could we have a `typedef` for the callback function? I think it will be much more readable and would allow documenting it. We could add something like this, alongside the type definitions in the beginning of the file:

```c
/**
 * wpe_view_backend_fullscreen_callback:
 * @userdata: ...
 * @enable: ...
 * 
 * Type of functions used by an embedder to implement fullscreening web views.
 *
 * Returns: Whether the operation was completed.
 */
typedef bool (*wpe_view_backend_fullscreen_callback)(void *userdata, bool enable);
```

and then here, change the declaration of the function to:

```suggestion
wpe_view_backend_set_platform_fullscreen_handler(struct wpe_view_backend*,
        wpe_view_backend_fullscreen_callback callback, void* userdata);
```

**Review by aperezdc — CHANGES_REQUESTED** [review]
This is starting to look ready for merging, thanks for working on it! 💪🏼 

I have left a few comments with suggestions, and it would be great if we could have @zdobersek and/or @carlosgcampos check this and rubber-stamp that they are okay with the new API.

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done, thanks for the documentation example!

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h:243]: Done, thanks for the documentation example!

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done

File: src/view-backend.c
Comment (MortimerGoro) [reply src/view-backend.c]: Done. I normally use NULL (or nullptr in C++11) but followed the style of the other setters which were using 0.

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done, thanks for the example!

**Review by aperezdc — CHANGES_REQUESTED** [review]
This looks almost ready, thanks! It will only need a couple of touch ups in the documentation comments.

In the meantime, I will try to get someone else to do a review as well, I would prefer to have one more person agree on new API before merging.

As a side note, I noticed a bug in HotDoc while trying to build the documentation added by this MR locally, and I have reported it: https://github.com/hotdoc/hotdoc/issues/230 🙃

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done. Your explanations were great, I missed filling that in this comment :)

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: done

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done

**Review by aperezdc — COMMENTED** [review]
This looks mergeable to me. Waiting to have a second opinion on the new API, hopefully @zdobersek and/or @carlosgcampos can take a look soon 👍🏼

File: include/wpe/view-backend.h
Comment (carlosgcampos) [inline include/wpe/view-backend.h:147]: Same for platform here, just set_fullscreen_handler

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Thanks for the review!

Done. I agree with the change, the new documentation already specifies what it can be used for.

File: include/wpe/view-backend.h
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done

**Review by aperezdc — APPROVED** [review]
Thanks @carlosgcampos for commenting, I think we can consider this now approved, right?

**Comment by aperezdc** [comment]
Let's merge this, I would like to make a development release including this feature, and it looks like @carlosgcampos should be happy now with the function name changes applied. If there are any other tweaks on top of this we may want before 1.12, we can use follow-up PRs.

**Pull request merged**
